rust/src/normalization.rs

```diff
@@ -120,6 +121,294 @@ pub(crate) fn rewrite_rpmdb_timestamps<F: Read + Write + Seek>(rpmdb: &mut F) ->
+        if line.is_empty() {
+            continue;
+        }
+
+        let mut bits = line.split(char::is_whitespace).filter(|s| !s.is_empty());
+        if let Some(m) = bits.next() {
+            if m == "%_db_backend" {
+                db_backend = bits.last().map(|s| s.to_string());
+                break;
+            }
+        };
+    }
+
+    let db_backend = db_backend.ok_or_else(|| anyhow!("Unable to determine rpmdb backend"))?;
+
+    match db_backend.as_str() {
+        "bdb" | "b-d-b" => rpmdb_bdb::normalize_rpmdb_bdb(rootfs, rpmdb_path, source_date),
+        "ndb" => Ok(()),
+        "sqlite" => Ok(()),
+        _ => Err(anyhow!("Unknown rpmdb backend: {}", db_backend)),
+    }
+}
+
+mod rpmdb_bdb {
+    use anyhow::{anyhow, Context, Result};
+    use binread::{BinRead, BinReaderExt};
+    use lazy_static::lazy_static;
+    use openat::SimpleType;
+    use sha2::Digest;
+    use std::ffi::OsStr;
+    use std::io::{Read, Seek, SeekFrom, Write};
+    use std::os::unix::io::AsRawFd;
+    use std::path::{Path, PathBuf};
+    use subprocess::{Exec, Redirection};
+
+    #[derive(BinRead, Debug, Clone, Copy, PartialEq, Eq)]
+    #[repr(u8)]
+    #[br(repr=u8)]
+    enum PageType {
+        IBTree = 3,
+        LBTree = 5,
+        Overflow = 7,
+        HashMeta = 8,
+        BTreeMeta = 9,
+        Hash = 13,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(little)]
+    struct MetaHeader {
+        lsn: u64,
+        pgno: u32,
+        magic: u32,
+        version: u32,
+        pagesize: u32,
+        encrypt_alg: u8,
+        page_type: PageType,
+        metaflags: u8,
+        unused1: u8,
+        free: u32,
+        last_pgno: u32,
+        nparts: u32,
+        key_count: u32,
+        record_count: u32,
+        flags: u32,
+        uid: [u8; 20],
+    }
+
+    const BTREE_MAGIC: u32 = 0x00053162;
+    const HASH_MAGIC: u32 = 0x00061561;
+    const PAGE_HEADER_SIZE: u16 = 26;
+    const PAGE_HEADER_MAGIC_OFFSET: u64 = 0x34;
+
+    #[derive(BinRead, Debug)]
+    #[br(little)]
+    struct PageHeader {
+        lsn: u64,
+        pgno: u32,
+        prev_pgno: u32,
+        next_pgno: u32,
+        entries: u16,
+        hf_offset: u16,
+        level: u8,
+        page_type: PageType,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(repr=u8)]
+    #[repr(u8)]
+    enum BTreeItemType {
+        KeyData = 1,
+        Duplicate = 2,
+        Overflow = 3,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(little)]
+    struct BTreeItem {
+        len: u16,
+        item_type: BTreeItemType,
+    }
+
+    #[derive(BinRead, Debug)]
+    #[br(repr=u8)]
+    #[repr(u8)]
+    enum HashItemType {
+        KeyData = 1,
+        Duplicate = 2,
+        Offpage = 3,
+        OffDup = 4,
+    }
+
+    lazy_static! {
+        static ref PROC_SELF_CWD: PathBuf = PathBuf::from("/proc/self/cwd");
+        static ref PROC_SELF_FD: PathBuf = PathBuf::from("/proc/self/fd");
+    }
+
+    pub(super) fn normalize_rpmdb_bdb<P: AsRef<Path>>(
```

Comment:
There's actually nothing specific to rpm about this, correct?

I wonder if it's worth calling this `bdb_normalization` and structuring it a bit more as if it might be a new crate in the future - exposed via a small `bdb-normalize-rs` CLI binary too.

Among other things, given the importance (IMO) of the "random process data ends up in the file" problem, it seems like people would want to do this even outside of rpm-ostree.  At least for Fedora derivatives, this seems quite useful for the container base image.

OTOH, I guess this problem will (hopefully!) slowly solve itself as the sqlite backend progresses through even into LTS distributions.